### PR TITLE
ddl: Refine the error message of `add index`, to avoid error overwritten

### PR DESCRIPTION
**Pull request by tsthght**
### What problem does this PR solve?

UCP #15152

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

### CODE REVIEW

File: ddl/rollingback.go [hunk 1]
```diff
@@ -311,7 +311,11 @@ func convertJob2RollbackJob(w *worker, d *ddlCtx, t *meta.Meta, job *model.Job)
 			logutil.Logger(w.logCtx).Info("[ddl] the DDL job is cancelled normally", zap.String("job", job.String()), zap.Error(err))
 		}
 
-		job.Error = toTError(err)
+		if job.Error == nil {
+			job.Error = toTError(err)
+		} else {
+			job.Error = toTError(job.Error.GenWithStack("[%s] %s", toTError(err).ToSQLError().Message, job.Error.ToSQLError().Message))
```
Comment (zimulala) [inline ddl/rollingback.go]: "current error msg:  %s, original error msg: %s"

Besides, if we encounter an error in the function of "rollingbackXxxx", we may get the error `current error msg:  err1, original error msg: err0`. Then we rolling back the DDL job again. In the end, we may get the error like `current error msg:  err2, original error msg: current error msg:  err1, original error msg: err0`. 
Perhaps we can confirm that the initial error message is not overwritten at the time of rolling back, and that this error format is handled only when the rollingback ends.

Comment (tsthght) [reply ddl/rollingback.go]: Thanks, I had modified it. PTAL @zimulala

**Comment by tangenta** [comment]
The objective of issue #15152 is to enrich the error message like `[ddl:8214]Cancelled DDL job`, isn't it?

https://github.com/pingcap/tidb/blob/82829e1fa5b908cf4b6ea8c97ba14a38fcde81fa/ddl/serial_test.go#L452

**Comment by zimulala** [comment]
@tsthght 
Please address the comment.

File: ddl/ddl.go [hunk 2]
```diff
@@ -491,6 +491,11 @@ func (d *ddl) doDDLJob(ctx sessionctx.Context, job *model.Job) error {
 		}
 
 		if historyJob.Error != nil {
+			if historyJob.State == model.JobStateRollbackDone && !historyJob.Error.Equal(errCancelledDDLJob) {
```
Comment (zimulala) [inline ddl/ddl.go:494]: Why add the condition of `!historyJob.Error.Equal(errCancelledDDLJob)`?

Comment (tsthght) [reply ddl/ddl.go:494]: Does cancel ddl job will also return the error?

Comment (tsthght) [reply ddl/ddl.go]: Thanks, I had changed it~

File: ddl/ddl.go [hunk 3]
```diff
@@ -491,6 +491,11 @@ func (d *ddl) doDDLJob(ctx sessionctx.Context, job *model.Job) error {
 		}
 
 		if historyJob.Error != nil {
+			if historyJob.State == model.JobStateRollbackDone || historyJob.State == model.JobStateCancelled {
+				historyJob.Error = historyJob.Error.Class().New(historyJob.Error.Code(),
```
Comment (zimulala) [inline ddl/ddl.go:495]: We'd better add a check if `historyJob.Error` is equal to `errCancelledDDLJob`. Then it needs not to return the same error message twice.

Comment (tsthght) [reply ddl/ddl.go:538]: Thanks, I had modified it.  @zimulala  PTAL

**Pull request merged**